### PR TITLE
CONTRACTS: add function-contract mapping to the CLI

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
Enables checking or replacing a contract against a function of a different name as long as signatures are compatible, using :

```
--enforce-contract <function>/<contract>
--enforce-contract-rec <function>/<contract>
--replace-call-with-contract <function>/<contract>
```

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/contracts-dfcc/function-contract-mapping/test-enforce-warning-not-found.desc [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+main.c
+--malloc-may-fail --malloc-fail-null --dfcc main --enforce-contract foo/my_contractt _ --pointer-check --pointer-primitive-check --pointer-overflow-check
+^Contract 'my_contractt' not found, deriving empty pure contract 'contract::my_contractt' from function 'foo'$
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+--
+--
+Checks that we get a warning that a default contract is derived from the
+signature of the function being checked when specifying a non-existing contract.
```
Comment (tautschnig) [inline regression/contracts-dfcc/function-contract-mapping/test-enforce-warning-not-found.desc]: No newline at end of file.

File: src/goto-instrument/contracts/contracts.h [hunk 2]
```diff
@@ -29,19 +29,22 @@ Date: February 2016
...
-  "                              replace calls to fun with fun's contract\n"
+  " --replace-call-with-contract <function>[/contract]\n"                      \
+  "                              replace calls to function with contract\n"
 
 #define FLAG_ENFORCE_CONTRACT "enforce-contract"
 #define HELP_ENFORCE_CONTRACT                                                  \
-  " --enforce-contract <fun>     wrap fun with an assertion of its contract\n"
+  " --enforce-contract <function>[/contract]"                                  \
+  "                              wrap function with an assertion of contract\n"
+// clang-format on
```
Comment (tautschnig) [inline src/goto-instrument/contracts/contracts.h:47]: The pull-request description mentions `--enforce-contract-rec`?!

File: src/goto-instrument/contracts/dynamic-frames/dfcc.cpp [hunk 3]
```diff
@@ -47,6 +47,66 @@ Author: Remi Delmas, delmarsd@amazon.com
...
+  auto const function_name_end = cli_flag_str.find('/');
+  if(function_name_end == std::string::npos)
+  {
+    // <function_name> -> (<function_name>, <function_name>)
+    return std::make_pair(cli_flag, cli_flag);
+  }
+
+  // <function_name>/<contract_name> -> (<function_name>, <contract_name>)
+  std::string function_name = cli_flag_str.substr(0, function_name_end);
+  std::string contract_name = cli_flag_str.substr(function_name_end + 1);
```
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc.cpp]: You might use `split_string(cli_flag_str, '/')` and then inspect the resulting vector.

File: src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.cpp [hunk 4]
```diff
@@ -110,39 +110,58 @@ void dfcc_contract_handlert::add_contract_instructions(
...
     new_symbol.pretty_name = pure_contract_id;
     new_symbol.is_property = true;
-    new_symbol.type = contract_symbol.type;
-    new_symbol.mode = contract_symbol.mode;
-    new_symbol.module = contract_symbol.module;
-    new_symbol.location = contract_symbol.location;
+    new_symbol.type = function_symbol.type;
+    new_symbol.mode = function_symbol.mode;
+    new_symbol.module = function_symbol.module;
+    new_symbol.location = function_symbol.location;
```
Comment (tautschnig) [inline src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.cpp]: Can you also change this code to make use of a non-trivial `symbolt` constructor that's now available?

**Review by tautschnig — APPROVED** [review]

**Review by feliperodri — APPROVED** [review]
Modulo Michael's comments.

File: src/goto-instrument/contracts/contracts.h (hunk 2 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.h:47]: yes, the option is declared in `dfcc.h`

File: src/goto-instrument/contracts/dynamic-frames/dfcc.h [hunk 5]
```diff
@@ -56,19 +57,32 @@ class optionst;
...
-// clang-format on
+  "--dfcc <harness>   activate dynamic frame condition checking for function\n"\
+  "                   contracts using the given harness as entry point"
 
-// clang-format off
 #define FLAG_ENFORCE_CONTRACT_REC "enforce-contract-rec"
 #define OPT_ENFORCE_CONTRACT_REC "(" FLAG_ENFORCE_CONTRACT_REC "):"
 #define HELP_ENFORCE_CONTRACT_REC                                              \
-  " --enforce-contract-rec <fun>  wrap fun with an assertion of its contract\n"\
+  " --enforce-contract-rec <function>[/<contract>]"                            \
```
Comment (remi-delmas-3000) [inline src/goto-instrument/contracts/dynamic-frames/dfcc.h:66]: @tautschnig  here it is

File: src/goto-instrument/contracts/dynamic-frames/dfcc.cpp (hunk 3 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc.cpp]: fixed

File: src/goto-instrument/contracts/contracts.h (hunk 2 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/contracts.h:47]: fixed

File: regression/contracts-dfcc/function-contract-mapping/test-enforce-warning-not-found.desc (hunk 1 above)
Comment (remi-delmas-3000) [reply regression/contracts-dfcc/function-contract-mapping/test-enforce-warning-not-found.desc]: fixed

File: src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.cpp (hunk 4 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/contracts/dynamic-frames/dfcc_contract_handler.cpp]: now using symbolt{name, type, mode}

**Pull request merged**